pallets/asset/src/lib.rs

```diff
@@ -1945,116 +1966,45 @@ impl<T: Config> Module<T> {
-        Self::ensure_asset_name_bounded(&name)?;
-        if let Some(fr) = &funding_round {
-            Self::ensure_funding_round_name_bounded(fr)?;
-        }
-        Self::ensure_asset_idents_valid(&identifiers)?;
-        Self::ensure_asset_type_valid(asset_type)?;
-
-        Self::ensure_create_asset_parameters(&ticker)?;
-
-        // Ensure its registered by DID or at least expired, thus available.
-        let available = match Self::is_ticker_available_or_registered_to(&ticker, did) {
-            TickerRegistrationStatus::RegisteredByOther => {
-                fail!(Error::<T>::TickerAlreadyRegistered)
-            }
-            TickerRegistrationStatus::RegisteredByDid => false,
-            TickerRegistrationStatus::Available => true,
-        };
-
-        // If `ticker` isn't registered, it will be, so ensure it is fully ascii.
-        if available {
-            Self::verify_ticker_characters(&ticker)?;
-        }
-
-        let token_did = Identity::<T>::get_token_did(&ticker)?;
-        // Ensure there's no pre-existing entry for the DID.
-        // This should never happen, but let's be defensive here.
-        Identity::<T>::ensure_no_id_record(token_did)?;
-
-        // Ensure that the caller has relevant portfolio permissions
-        let user_default_portfolio = PortfolioId::default_portfolio(did);
-        Portfolio::<T>::ensure_portfolio_custody_and_permission(
-            user_default_portfolio,
-            did,
-            secondary_key.as_ref(),
-        )?;
-
-        // Charge protocol fees.
-        T::ProtocolFee::charge_fees(&{
-            let mut fees = ArrayVec::<_, 2>::new();
-            if available {
-                fees.push(ProtocolOp::AssetRegisterTicker);
-                fees.push(ProtocolOp::AssetCreateAsset);
-            }
-            fees
-        })?;
-
-        //==========================================================================
-        // At this point all checks have been made; **only** storage changes follow!
-        //==========================================================================
+    ) -> DispatchResult {
+        T::ProtocolFee::charge_fee(ProtocolOp::AssetCreateAsset)?;
+        Self::unverified_register_ticker(ticker, caller_did, None)?;
 
         Identity::<T>::commit_token_did(token_did, ticker);
+        let token = SecurityToken::new(Zero::zero(), caller_did, divisible, asset_type);
+        Tokens::insert(ticker, token);
 
-        // Register the ticker or finish its registration.
-        if available {
-            // Ticker not registered by anyone (or registry expired), so register.
-            Self::unverified_register_ticker(&ticker, did, None);
-        } else {
-            // Ticker already registered by the user.
-            <Tickers<T>>::mutate(&ticker, |tr| {
-                if let Some(tr) = tr {
-                    tr.expiry = None;
-                }
-            });
+        AssetNames::insert(ticker, asset_name.clone());
```

Comment:
Done need clone here, can pass by reference.
```suggestion
        AssetNames::insert(ticker, &asset_name);
```